spec-cordova/prepare.spec.js

```diff
@@ -17,175 +17,265 @@
+                    done();
+                });
+            });
         });
-        it('Test 002 : should not run inside a cordova-based project with no platforms', function(done) {
-            list_platforms.and.returnValue([]);
-            prepare().then(function() {
-                expect('this call').toBe('fail');
-            }, function(err) {
-                expect('' + err).toContain('No platforms added to this project. Please use `cordova platform add <platform>`.');
-            }).fin(done);
+
+        describe('success', function () {
+            beforeEach(function () {
+                spyOn(util, 'cdProjectRoot').and.returnValue(project_dir);
+                spyOn(util, 'preProcessOptions').and.callFake(function(options) {
+                    let platforms = options.platforms || [];
+                    return {'platforms':platforms};
+                });
+            });
+            it('should fire the before_prepare hook and provide platform and path information as arguments', function(done) {
+                prepare({}).then(function() {
+                    expect(HooksRunner.prototype.fire.calls.argsFor(0)[0]).toEqual('before_prepare');
+                    done();
+                });
+            });
+            it('should invoke restore module\'s installPlatformsFromConfigXML method', function(done) {
+                prepare({}).then(function() {
+                    expect(restore.installPlatformsFromConfigXML).toHaveBeenCalled();
+                    done();
+                }).fail(function(err){
+                    expect(err).toBeUndefined();
+                    done();
+                });
+            });
+            it('should retrieve PlatformApi instances for each platform provided', function(done) {
+                prepare({'platforms':['android', 'ios']}).then(function() {
+                    expect(platforms.getPlatformApi.calls.count()).toBe(4);
+                    expect(platforms.getPlatformApi.calls.argsFor(0)[0]).toBe('android');
+                    expect(platforms.getPlatformApi.calls.argsFor(0)[1]).toBe('/some/path/platforms/android');
+                    expect(platforms.getPlatformApi.calls.argsFor(1)[0]).toBe('ios');
+                    expect(platforms.getPlatformApi.calls.argsFor(1)[1]).toBe('/some/path/platforms/ios');
+                    expect(platforms.getPlatformApi.calls.argsFor(2)[0]).toBe('android');
+                    expect(platforms.getPlatformApi.calls.argsFor(3)[0]).toBe('ios');
+                    done();
+                }).fail(function(err){
+                    expect(err).toBeUndefined();
+                    done();
+                });
+            });
+            it('should invoke restore module\'s installPluginsFromConfigXML method', function(done) {
+                prepare({platforms:[]}).then(function() {
+                    expect(restore.installPluginsFromConfigXML).toHaveBeenCalled();
+                    done();
```

Comment:
For these kind of tests, where we want to set expectations on something happening in _either_ the success or failure handler, I like to structure the test like so:

```
prepare().then(function () {
    expect(something).toBe(true);
}).fail(function (e) {
    fail('unexpected failure handler invoked');
    console.error(e);
}).done(done);
```

The reasons I do it this way are:
1. If the handler that I am not expecting gets invoked, the test fails immediately. With your implementation above, we'd have to wait for the timeout to kick in (which is by default 5 seconds IIRC). That can add up if a lot of tests are exhibiting that behaviour.
2. There's always one spot to call `done`, using the final promise resolution handler (either `done` or `fin` - both should be identical).
3. If the wrong handler gets invoked, by logging out the error, the person that sees the error can immediately start debugging, as `console.log` or `console.error` can handle Error object stack traces and print them out too. Maybe there's a way to combine the `fail` + `console.error` calls into one. I know if you do `expect(err).toBeUndefined()`, you will only get the error message and not the stack trace, which makes debugging harder. Perhaps `fail('unexpected fail handler invoked with error: ' + err)` would work too?